### ISSUE TITLE
notation for N- and C-terminal modifications

### ISSUE THREAD

**Issue by lars20070** [OpenMS/OpenMS#1792]
We could use a notation which represents N- and C-termini explicitly. For example

```
nX ... Yc
```

A modification `(m)` can occur on the termini or side chains.

```
n(m)X ... Yc
nX(m) ... Yc
nX ... Y(m)c
nX ... Yc(m)
```

In our current notation, we cannot distinguish between the last two cases.

```
(m)X ... Y
X(m) ... Y
X ... Y(m)
X ... Y(m)
```

In addition, the new notation simplifies the search for terminal modifications by looking for substrings `n(m)` or `c(m)` in the amino acid sequence.

Should we  denote the termini explicitly?

**Comment by timosachsenberg** [comment]
In general I am supporting an encoding that allows to distinguish the termini. 
Don't know if we should use n and c as both could be converted to amino acids. Maybe just use "."
if a terminal modification is present that is not attached to the amino acid.
XXXXXXXX.(MOD) and XXXXXXXX(MOD) could then be disstinguished

for back compatibility reasons I would also allow both:

(MOD).XXXXXXXX and (MOD)XXXXXXXX for denoting a N-term mod.

**Comment by lars20070** [comment]
+1 for the notation.

-1 for the back compatibility. An advantage of the new notation is that we can simply search for substrings `(MOD).`and `.(MOD)`.

**Comment by timosachsenberg** [comment]
I am talking about parsing the old notation - maybe there was a misconception

**Comment by mwalzer** [comment]
Internally, terminal mods are kept separate [anyways](http://ftp.mi.fu-berlin.de/pub/OpenMS/documentation/html/classOpenMS_1_1AASequence.html). So this would mean the change of the toString() only?

**Comment by timosachsenberg** [comment]
yes basically how we parse and write out (toString())

**Comment by mwalzer** [comment]
maybe leave the toString as is and introduce a 'prettyString' function that will do such extra annotations for compatib. reasons

**Comment by timosachsenberg** [comment]
yeah maybe .. or maybe not ;) Both come with some caveats...
e.g. everything that is generated by prettyString will not be downwards compatible so the dev. must decide if he wants or not wants to be backward compatible - a question he most likely can't answer without deep analysis of all implications

**Comment by hendrikweisser** [comment]
Regarding "prettyString": How about a flag in "toString" instead, that defaults to the new notation? I support the one with dots for the ends.

**Comment by timosachsenberg** [comment]
I like the flag idea because it makes deprecating and abandoning the old format easier in the future.

**Comment by mwalzer** [comment]
> I support the one with dots for the ends.

Sure, me too - but do all our tools and workflows? 
> How about a flag in "toString"

Excellent idea :+1:

**Comment by hroest** [comment]
any idea how other people encode N terminal modifications? What does TPP do? Just as we dont want to re-invent the wheel here

**Comment by lars20070** [comment]
TPP is using the `nX ... Yc` notation.

**Comment by timosachsenberg** [comment]
also note that . is in many publications or software used in a very similar context: if e.g. flanking amino acids are encoded as in: M.PEPTIDEPEPTIDER.Y
in a way, here "." is already used to encode the termini.
I don't like the "n" and "c" notation of TPP very much as lower case letters are often used to denote modified amino acids.

**Comment by cbielow** [comment]
MaxQuant is using _ .. so very similar to what Timo suggested.
Strangely, they write _(ac)AAA... for an n-terminal acetylation. Should be (ac)_AAA...?!

**Comment by cbielow** [comment]
Can we estimate how much work it is to completely switch to the new notation?!
I'm a little worried that having two notations will confuse developers and users, because if something crashes (or worse: does not crash) and only the new notation is in place, then it's pretty clear what needs to be done.

**Comment by lars20070** [comment]
@cbielow IMO, the MaxQuant notation makes sense. The dots represent the termini and are not a divider. The mods are always to the right of the amino acid or terminus.
```
.(m)X ... Y.
.X(m) ... Y.
.X ... Y(m).
.X ... Y.(m)
```

**Comment by cbielow** [comment]
ahh.. early in the morning ... :)
Thx.

**Comment by lars20070** [comment]
I am for a complete switch to the new dot notation, i.e. no flag in `toString()` [1]. The old notation is  ambiguous and should not be used in future. But `fromString()` [2] should allow for both old and new notation, i.e. even strings without dots are permitted. In this way, even old `idXML` without dots can be read. `X ... Y(m)` would be interpreted as C-terminal modification i.e.  `X ... Y.(m)`.

Regarding the effort to make the transition, I hope only [1] and [2] need to be modified.

[1] https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/CHEMISTRY/AASequence.cpp#L108
[2] https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/CHEMISTRY/AASequence.cpp#L1030

**Comment by timosachsenberg** [comment]
A direct switch is also ok for me.
Thinking about it again, introducing the flag does not really help in this regards. It only allows users that code in C++/pyOpenMS to generate the old strings. As we have a rather limited developers base we could communicate the switch via our mailing list. Adoption of existing scripts would probably only involve removing every occurrence of ".".

**Comment by hendrikweisser** [comment]
> I am for a complete switch to the new dot notation, i.e. no flag in toString()

One thing to consider here is how often we have to create "old-style" amino acid strings in OpenMS, e.g. to generate different file formats (mzid?). Of course it's possible to create a string with dots at the ends and then remove the dots, but if that happens often enough in the code base, having a flag would simplify things.

**Comment by timosachsenberg** [comment]
I would guess that to unmodifiedString would then be the choice as it does not require the .

**Comment by lars20070** [comment]
+1 @timosachsenberg

**Comment by mwalzer** [comment]
Yeah, real identification formats do not encode that information, rather keep it in a structured though separate way, so using ```toUnmodifiedString``` and then registering modifications at the designated places. Leaves less leeway for misunderstandings :wink:

**Comment by timosachsenberg** [comment]
https://github.com/timosachsenberg/OpenMS/tree/feature/new_sequence_notation
for a quick implementation. many tests are failing (as expected).

**Comment by lars20070** [comment]
Never seen mods in square brackets in OpenMS. What's that code for?
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/CHEMISTRY/AASequence.cpp#L787

**Comment by hendrikweisser** [comment]
> Never seen mods in square brackets in OpenMS.

Square brackets are used to specify mods via mass shifts.

**Comment by lars20070** [comment]
Thanks @hendrikweisser. - Which tool is using this notation? Is there a flag?

**Comment by hroest** [comment]
TPP uses square bracket notation.  It is one of the most common cases.

sent from my smartphone -- you may keep any typos you find.

**Comment by lars20070** [comment]
Resolved in #1970.
